src/vmware/vmxnet3_net.c

```diff
@@ -0,0 +1,581 @@
+
+    dev->rx_pbuf[rid][idx] = (struct pbuf*)x;
+
+    rxd->addr = physical_from_virtual(x+1);
+    rxd->len = vdev->rxbuflen;
+
+    rxd->btype = (rid == 0 && (idx % 1) == 0) ? VMXNET3_BTYPE_HEAD : VMXNET3_BTYPE_BODY;
+    rxd->gen = rxr->vxrxr_gen;
+
+    if (++rxr->vxrxr_refill_start == rxr->vxrxr_ndesc) {
+        rxr->vxrxr_refill_start = 0;
+        rxr->vxrxr_gen ^= 1;
+    }
+}
+
+void
+vmxnet3_receive(vmxnet3 vdev)
+{
+    vmxnet3_pci dev = vdev->dev;
+    struct vmxnet3_rxqueue *rxq = dev->vmx_rxq[0];
+    struct vmxnet3_comp_ring *rxc = &rxq->vxrxq_comp_ring;
+
+    for(;;) {
+        struct vmxnet3_rxcompdesc *rxcd = &rxc->vxcr_u.rxcd[rxc->vxcr_next];
+
+        if (rxcd->gen != rxc->vxcr_gen)
+            break;
+        read_barrier();
+
+        assert(rxcd->qid <= 2);
+
+        if (++rxc->vxcr_next == rxc->vxcr_ndesc) {
+            rxc->vxcr_next = 0;
+            rxc->vxcr_gen ^= 1;
+        }
+
+        u32 rid = rxcd->qid;
+        u32 idx = rxcd->rxd_idx;
+        u32 length = rxcd->len;
+        struct vmxnet3_rxring *rxr = &rxq->vxrxq_cmd_ring[rid];
+        struct vmxnet3_rxdesc *rxd = &rxr->vxrxr_rxd[idx];
+        struct pbuf *m = dev->rx_pbuf[rid][idx];
+
+        assert(m != NULL);
+
+        while(rxr->vxrxr_refill_start != idx) {
+            rxd[rxr->vxrxr_refill_start].gen = rxr->vxrxr_gen;
+            if (++rxr->vxrxr_refill_start == rxr->vxrxr_ndesc) {
+                rxr->vxrxr_refill_start = 0;
+                rxr->vxrxr_gen ^= 1;
+            }
+        }
+
+        if (rxcd->eop && rxcd->error) {
+            vmxnet3_discard(dev, rid, idx);
+            goto next;
+        }
+
+        /* Check and handle SOP/EOP state errors */
+        if (rxcd->sop && dev->currpkt_head) {
+            receive_buffer_release(dev->currpkt_head);
+            dev->currpkt_head = dev->currpkt_tail =  NULL;
+//            stats.rx_drops++;
+        } else if (!rxcd->sop && !dev->currpkt_head) {
+//            stats.rx_drops++;
+            vmxnet3_discard(dev, rid, idx);
+            goto next;
+        }
+
+       if (rxcd->sop) {
+            assert(rxd->btype == VMXNET3_BTYPE_HEAD);
+            assert((idx % 1) == 0);
+            assert(dev->currpkt_head == NULL);
+
+            if (length == 0) {
+                vmxnet3_discard(dev, rid, idx);
+                goto next;
+            }
+
+            vmxnet3_newbuf(vdev, rid);
+
+            m->tot_len = length;
+            m->len = length;
+
+            dev->currpkt_head = dev->currpkt_tail = m;
+        } else {
+            assert(rxd->btype == VMXNET3_BTYPE_BODY);
+            assert(dev->currpkt_head != NULL);
+
+            vmxnet3_newbuf(vdev, rid);
+
+            m->len = length;
+            dev->currpkt_head->tot_len += length;
+            dev->currpkt_tail->next = m;
+            dev->currpkt_tail = m;
+        }
+
+        if (rxcd->eop) {
+            if (vdev->n->input(dev->currpkt_head, vdev->n) != ERR_OK) {
```

Comment:
lwIP input needs to happen in bottom half processing, not at interrupt level. For virtio devices such deferral is taken care of automatically by virtqueue. You may want to queue up work for bh processing here ... see xennet_service_rx_ring() (irq portion) and xennet_rx_service_bh() (thunk run on bhqueue) in src/xen/xennet.c for an example.